File: homework2/steven_law/q5_Stack.cpp
// Time spent: 20 mins 51 seconds

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* next;
};

class Stack {
private:
    Node* head;

public:
    Stack() : head(nullptr) {}
    // returns the top item in the stack. O(1) time.
    int top() {
        if(head == nullptr) return -1;
        return head ->data;
    }

    // adds x to the top of the stack. O(1) time.
    void push(int x) {
        Node* newNode = new Node();
        newNode->data = x;
        newNode->next = head;
        head = newNode;
         
    }

    // removes and returns the top item in the stack. O(1) time.
    int pop() {
        if(head ==nullptr) return -1;

        int val = head->data;

        Node* temp = head;
        head = head->next; 
        delete temp;

        return val;

    }

    // returns a boolean indicating whether the stack is empty. O(1) time.
    bool isEmpty() {
        if(head == nullptr) return true;
        return false;
    }
};

int main() {

    return 0;
}

File: homework2/saad_amin/q5_Stack.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class Stack:
    def __init__(self):
        self.head = None 
        
    def top(self):
        if self.head is None:
            return None
        return self.head.data
    
    def push(self, x):
        new_node = Node(x)
        new_node.next = self.head
        self.head = new_node
        
    def pop(self):
        if self.head is None:
            return None

        val = self.head.data
        self.head = self.head.next
        return val
    
    def isEmpty(self):
        return self.head is None